packages/SwingSet/misc-tools/vat-map-from-swingstore.js

```diff
@@ -0,0 +1,155 @@
+const vdidToBundleID = vdid => get(`${deviceVatAdmin}.vs.slot.${vdid}`);
+const kdidToVdid = kdid => get(`${deviceVatAdmin}.c.${kdid}`);
+const kdidToBundleID = kdid => vdidToBundleID(kdidToVdid(kdid));
+
+const kdNumber = vatID => Number(vatID.slice(1)); // 'v12' -> 12
+const importedKdids = vatID => {
+  const kdids = [];
+  // read all vNN.c.d-NN keys, the values are kdNN
+  const start = `${vatID}.c.d-`;
+  const end = `${vatID}.c.d.`;
+  for (const row of sqlGetRange.all(start, end)) {
+    kdids.push(row.value);
+  }
+  sortWith(kdids, kdNumber);
+  return kdids;
+};
+const lastImportedKdid = vatID => importedKdids(vatID).at(-1);
+
+const output = { runID, currentZCF: undefined, vats: {} };
+
+let zoeVatID;
+for (const vatID of allVatIDs) {
+  const source = getJSON(`${vatID}.source`).bundleID;
+  const options = getJSON(`${vatID}.options`);
+  const { name, critical, workerOptions } = options;
+  if (name === 'zoe') {
+    zoeVatID = vatID;
+  }
+  const { type } = workerOptions;
+  let lockdownBundleID;
+  let supervisorBundleID;
+  if (type === 'xsnap') {
+    [lockdownBundleID, supervisorBundleID] = workerOptions.bundleIDs;
+  }
+
+  const { incarnation, endPos } = sqlGetSpan.get(vatID);
+
+  const vat = {
+    name,
+    critical,
+    incarnation,
+    endPos,
+    source,
+    type,
+    lockdownBundleID,
+    supervisorBundleID,
+  };
+  output.vats[vatID] = vat;
+
+  console.log(`${vatID}`);
+  console.log(` name: ${name}`);
+  console.log(` critical: ${critical}`);
+  console.log(` incarnation: ${incarnation}`);
+  console.log(` endPos: ${endPos}`);
+  console.log(` source: ${source}`);
+  console.log(` worker type: ${type}`);
+  console.log(` lockdown: ${lockdownBundleID}`);
+  console.log(` supervisor: ${supervisorBundleID}`);
+
+  // contract vats will launch from a ZCF bundle, and Zoe will label
+  // them as "zcf-*"
+  if (name.startsWith('zcf-')) {
+    if (!knownZCFIDs.includes(source)) {
+      console.log(`${vatID} claims to be contract but not using known ZCF`);
+      console.log(`TODO: maybe add to knownZCFIDs ?`);
+      console.log(`name: ${name}`);
+      console.log(`source: ${source}`);
+      process.exit(1);
+    }
+    const kdid = lastImportedKdid(vatID);
+    assert(kdid);
+    const contractBundleID = kdidToBundleID(kdid);
+    console.log(` contract bundle: ${contractBundleID}`);
+    vat.contractBundleID = contractBundleID;
+  }
+
+  console.log();
+}
+
+// figure out currently-available ZCF bundle by snooping v9-zoe's
+// baggage
+
+assert(zoeVatID);
+const zoeZCFBaggageName = 'zcfBundleCap';
+const zoeCurrentZCFvref = getJSON(`${zoeVatID}.vs.vc.1.s${zoeZCFBaggageName}`)
+  .slots[0];
+const zoeCurrentZCFkref = get(`${zoeVatID}.c.${zoeCurrentZCFvref}`);
+const zoeCurrentZCFBundleID = kdidToBundleID(zoeCurrentZCFkref);
+console.log(`current zoe ZCF: ${zoeCurrentZCFBundleID}`);
+output.currentZCF = zoeCurrentZCFBundleID;
+
+if (outputFn) {
+  fs.writeFileSync(outputFn, `${JSON.stringify(output)}\n`);
+}
```

Comment:
Suggestion: remove this parameter in favor of encouraging standard output redirection via `>`/`|`/etc.
```suggestion

console.log(JSON.stringify(output));
```
(and if so, replace each existing `console.log` with `console.warn`/`console.error` [which both print to standard error])

Reply:
I switched it to always console.log, but take an `--as-json` flag. So far I've run this tool a few times for the human-readable output, to find the vat I'm looking for, but then I wanted a machine-readable JSON format for other tooling, so this seems like a better compromise than needing to tease apart stdout vs stderr for the two forms.